core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java

```diff
@@ -231,4 +310,91 @@ public void stopSync(AlluxioURI uri) { }
   protected static String validatePath(String path) {
     return new AlluxioURI(path).toString();
   }
+
+  /**
+   * Represents a under filesystem operation.
+   */
+  private interface UfsOperation<T> {
+    /**
+     * Applies this operation.
+     *
+     * @return the result of this operation
+     */
+    T apply() throws IOException;
+  }
+
+  /**
+   * Represents a under filesystem operation with a boolean return type.
+   */
+  private interface UfsBooleanOperation<Boolean> {
+    /**
+     * Applies this operation.
+     *
+     * @return the result of this operation
+     */
+    boolean apply() throws IOException;
+  }
+
+  /**
+   * Retries the given under filesystem operation when it throws exceptions
+   * to resolve eventual consistency issue.
+   *
+   * @param op the under filesystem operation to retry
+   * @param description the description regarding the operation
+   * @return the operation result if operation succeed or returned true
+   */
+  private <T> T retryOnException(UfsOperation<T> op,
+      Supplier<String> description) throws IOException {
+    RetryPolicy retryPolicy = getRetryPolicy();
+    IOException thrownException = null;
+    while (retryPolicy.attempt()) {
+      try {
+        return op.apply();
+      } catch (IOException e) {
+        LOG.debug("{} attempt to {} failed with exception : {}", retryPolicy.getAttemptCount(),
+            description, e.getMessage());
+        thrownException = e;
+      }
+    }
+    throw thrownException;
+  }
+
+  /**
+   * Retries the given under filesystem operation when it throws exceptions or return false
+   * to resolve eventual consistency issue.
+   *
+   * @param op the under filesystem operation to retry
+   * @param description the description regarding the operation
+   * @return the operation result if operation succeed or returned true
+   */
+  private boolean retryOnFalse(UfsBooleanOperation<Boolean> op,
```

Comment:
can we write a helper method to combine the common logic between `retryOnException` and `retryOnFalse`? The only difference is what conditions they retry under.

I think the signature could be something like

```java
private <T> retry(Callable<T> op, BiFunction<Throwable, T, Boolean> shouldRetry)
```

Reply:
The `retryOnFalse` I have changed to only retry on false. Now the two functions have more different places, would you think it is still helpful to combine the logics?
```
  private boolean retryOnFalse(ObjectStoreOperation<Boolean> op,
      Supplier<String> description) throws IOException {
    RetryPolicy retryPolicy = getRetryPolicy();
    while (retryPolicy.attempt()) {
      if (op.apply()) {
        return true;
      } else {
        LOG.debug("Failed in attempt {} to {} ", retryPolicy.getAttemptCount(),
            description.get());
      }
    }
    return false;
  }

  private <T> T retryOnException(ObjectStoreOperation<T> op,
      Supplier<String> description) throws IOException {
    RetryPolicy retryPolicy = getRetryPolicy();
    IOException thrownException = null;
    while (retryPolicy.attempt()) {
      try {
        return op.apply();
      } catch (IOException e) {
        LOG.debug("Attempt {} to {} failed with exception : {}", retryPolicy.getAttemptCount(),
            description.get(), e.toString());
        thrownException = e;
      }
    }
    throw thrownException;
  }
```

Reply:
I think it would be a small improvement, but if you prefer two methods I'm fine keeping it how it is now.

FYI you can get java syntax highlighting in code snippets by starting the block with "```java"